File: backend/app/export/itw_export.py
"""Build the Iron Trainer Workout (.itw) intermediate representation.

Apple's native ``.workout`` file is an undocumented binary produced only on-device
by WorkoutKit's ``WorkoutComposition`` export — it cannot be generated on a server.
So the backend emits this neutral, versioned JSON instead. The iOS helper app
reads a ``.itw`` file and builds a real ``CustomWorkout`` / ``WorkoutPlan`` on the
device, then schedules it (using ``date``) or opens it in the Workout app.

The step shape is the same one ``fit_export``/``zwo_export`` consume
(``type``, ``duration_s``, ``distance_m``, ``notes``, ``target{type,low,high,unit}``),
so steps pass through unchanged.
"""

from __future__ import annotations

import json

SCHEMA_VERSION = 1

# Athlete thresholds carried in the file so it is self-contained and the iOS app
# can resolve relative targets without calling back to the server.
_THRESHOLD_KEYS = ("ftp", "threshold_hr", "max_hr", "threshold_pace_run", "css_swim")


def _workout_doc(workout: dict, athlete: dict) -> dict:
    """The per-workout .itw object. Each item is a standalone, decodable workout
    (carries its own schema_version) so it can be read singly or inside a plan."""
    return {
        "schema_version": SCHEMA_VERSION,
        "generator": "iron-trainer",
        "date": workout.get("date"),
        "sport": workout.get("sport"),
        "title": workout.get("title"),
        "description": workout.get("description"),
        "duration_s": workout.get("duration_s"),
        "distance_m": workout.get("distance_m"),
        "athlete": {k: athlete.get(k) for k in _THRESHOLD_KEYS},
        "steps": workout.get("steps") or [],
    }


def build_itw(workout: dict, athlete: dict | None = None) -> str:
    """Return the JSON string of a .itw file for one planned session."""
    return json.dumps(_workout_doc(workout, athlete or {}), ensure_ascii=False, indent=2)


def build_plan_itw(workouts: list[dict], plan: dict | None, athlete: dict | None = None) -> str:
    """Return the JSON string of a whole-plan .itw the iOS app fetches over HTTP:
    a top-level envelope plus a list of standalone per-workout docs."""
    athlete = athlete or {}
    plan_meta = None
    if plan:
        plan_meta = {
            "race_name": plan.get("race_name"),
            "race_date": plan.get("race_date"),
            "summary": plan.get("summary"),
        }
    doc = {
        "schema_version": SCHEMA_VERSION,
        "generator": "iron-trainer",
        "plan": plan_meta,
        "workouts": [_workout_doc(w, athlete) for w in workouts],
    }
    return json.dumps(doc, ensure_ascii=False, indent=2)

File: backend-v2/src/main/java/io/gamov/irontrainer/export/ItwExport.java
package io.gamov.irontrainer.export;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.gamov.irontrainer.athlete.Athlete;
import io.gamov.irontrainer.plan.PlannedWorkout;
import jakarta.enterprise.context.ApplicationScoped;

/** Port of app/export/itw_export.py — the neutral, versioned JSON the iOS app
 * turns into WorkoutKit compositions. Contract: schema_version 1, athlete
 * thresholds embedded (self-contained file), steps passed through verbatim
 * from structure_json. */
@ApplicationScoped
public class ItwExport {

    public static final int SCHEMA_VERSION = 1;

    private final ObjectMapper mapper = new ObjectMapper();

    public String workoutItw(PlannedWorkout w, Athlete a) {
        try {
            return mapper.writerWithDefaultPrettyPrinter().writeValueAsString(doc(w, a));
        } catch (Exception e) {
            throw new IllegalStateException("ITW serialization failed", e);
        }
    }

    private ObjectNode doc(PlannedWorkout w, Athlete a) throws Exception {
        ObjectNode doc = mapper.createObjectNode();
        doc.put("schema_version", SCHEMA_VERSION);
        doc.put("generator", "iron-trainer");
        doc.put("date", w.date);
        doc.put("sport", w.sport);
        doc.put("title", w.title);
        doc.put("description", w.description);
        if (w.durationS != null) doc.put("duration_s", w.durationS); else doc.putNull("duration_s");
        if (w.distanceM != null) doc.put("distance_m", w.distanceM); else doc.putNull("distance_m");
        ObjectNode athlete = doc.putObject("athlete");
        athlete.put("ftp", a == null ? null : a.ftp);
        athlete.put("threshold_hr", a == null ? null : a.thresholdHr);
        athlete.put("max_hr", a == null ? null : a.maxHr);
        athlete.put("threshold_pace_run", a == null ? null : a.thresholdPaceRun);
        athlete.put("css_swim", a == null ? null : a.cssSwim);
        // Steps pass through verbatim — same shape fit/zwo exporters consume.
        if (w.structureJson != null && !w.structureJson.isBlank()) {
            var parsed = mapper.readTree(w.structureJson);
            var steps = parsed.isArray() ? parsed : parsed.path("steps");
            doc.set("steps", steps.isArray() ? steps : mapper.createArrayNode());
        } else {
            doc.set("steps", mapper.createArrayNode());
        }
        return doc;
    }
}
